### PR TITLE
Duplicate experiment Key issue with multi feature flag

### PR DESCRIPTION
**Pull request by The-inside-man**
Summary
-------

-  Updated changes to use experiment_id_map
-  Build experiment_key_map from experiment_id_map
-  Update decision service to id instead of experiment key
-  Add test cases

Test plan
---------
-  FSC suite
Issues
------
- BugFix

### CODE REVIEW

File: tests/test_config.py [hunk 1]
```diff
@@ -41,6 +41,7 @@ def test_init(self):
                 self.config_dict['groups'][0]['trafficAllocation'],
             )
         }
+
```
Comment (keppel2) [inline tests/test_config.py:44]: ?

**Review by keppel2 — APPROVED** [review]

File: tests/test_config.py (hunk 1 above)
Comment (keppel2) [reply tests/test_config.py:44]: Only curious because it was by itself.

Comment (The-inside-man) [reply tests/test_config.py:44]: we need a new line to satisfy Flake8

Comment (The-inside-man) [reply tests/test_config.py:44]: formatter...

**Review by keppel2 — APPROVED** [review]

**Comment by The-inside-man** [comment]
Flake8 needs to be fixed... has both best practice and anti-pattern flagged so no winning:

./optimizely/project_config.py:577:17: W504 line break after binary operator
./optimizely/project_config.py:594:17: W503 line break before binary operator

**Review by Mat001 — APPROVED** [review]
LGTM. Good that FSC passes

**Comment by The-inside-man** [comment]
@msohailhussain If you can give the final OK, I can then merge to Master.

File: .flake8 [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [flake8]
 # E722 - do not use bare 'except'
-ignore = E722
+ignore = E722, W504
```
Comment (msohailhussain) [inline .flake8:3]: can you please explain in the comments, why W504 added.

File: optimizely/project_config.py [hunk 3]
```diff
@@ -557,3 +566,35 @@ def is_feature_experiment(self, experiment_id):
         """
 
         return experiment_id in self.experiment_feature_map
+
+    def get_variation_from_id_by_experiment_id(self, experiment_id, variation_id):
+        """ Gets experiment id and variation id
```
Comment (msohailhussain) [inline optimizely/project_config.py:571]: Gets only variation by variation id under specific experiment id.

Comment (msohailhussain) [reply optimizely/project_config.py:571]: please rephrase

File: optimizely/project_config.py [hunk 4]
```diff
@@ -557,3 +566,35 @@ def is_feature_experiment(self, experiment_id):
...
+                variation_id in self.variation_id_map_by_experiment_id[experiment_id]):
+            return self.variation_id_map_by_experiment_id[experiment_id][variation_id]
+
+        self.logger.error('Variation with id "%s" not defined in the datafile for experiment "%s".',
+                          variation_id, experiment_id)
+
+        return {}
+
+    def get_variation_from_key_by_experiment_id(self, experiment_id, variation_key):
+        """ Gets experiment id and variation key
```
Comment (msohailhussain) [inline optimizely/project_config.py:587]: please rephrase and have more clear explanation.

**Review by msohailhussain — APPROVED** [review]
lgtm. please address comments before merge and add an FSC link passing this PR.

File: .flake8 (hunk 2 above)
Comment (The-inside-man) [reply .flake8:6]: comment added explaining the reason for W504.

File: optimizely/project_config.py (hunk 4 above)
Comment (The-inside-man) [reply optimizely/project_config.py]: comments changed to better describe what is happening.

File: optimizely/project_config.py (hunk 3 above)
Comment (The-inside-man) [reply optimizely/project_config.py]: comments changed to better describe what is happening.

**Comment by The-inside-man** [comment]
Passing FSC - https://travis-ci.com/github/optimizely/fullstack-sdk-compatibility-suite/builds/232344312

**Comment by The-inside-man** [comment]
Link to FSC passing new test cases from "yasir/similar-rule-keys" : https://travis-ci.com/github/optimizely/fullstack-sdk-compatibility-suite/builds/232344687

**Comment by The-inside-man** [comment]
FSC Link for passing new cases: https://travis-ci.com/github/optimizely/fullstack-sdk-compatibility-suite/builds/232344687

**Pull request merged**